### ISSUE TITLE
convert frozon pb file to tflite but cannot find Output array which may be caused by that the frozen pb still have raining ops like Assign.

### ISSUE THREAD

**Issue by fzhu129** [tensorflow/tensorflow#20353]
I cannot generate tflite file which may be caused that frozon pb file still have training ops.

The operation like below:
use 
bazel-bin/tensorflow/python/tools/freeze_graph \
  --input_graph=/data/frankzhu/tmp/t2t_train/mnist/graph.pbtxt\
  --input_checkpoint=/data/frankzhu/tmp/t2t_train/mnist/model.ckpt-2000 \
  --input_binary=false \
  --output_graph=/data/frankzhu/tmp/mnist.pb \
  --output_node_names=save/restore_all

to get .tflite file

then use 

bazel run --config=opt \
  //tensorflow/contrib/lite/toco:toco -- \
  --input_file=/data/frankzhu/tmp/mnist.pb \
  --input_format=TENSORFLOW_GRAPHDEF \
  --output_format=TFLITE \
  --output_file=/data/frankzhu/tmp/mnist_q8.tflite \
  --inference_type=QUANTIZED_UINT8 \
  --input_type=FLOAT \
  --input_arrays=input \
  --output_arrays=save/restore_all/NoOp --input_shapes=1,28,28,1

But get error:
018-06-11 09:42:39.760033: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Assign
2018-06-11 09:42:39.760052: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1366] Converting unsupported operation: Assign
2018-06-11 09:42:39.774575: F tensorflow/contrib/lite/toco/tooling_util.cc:806] Check failed: model.HasArray(output_array) Output array not found: save/restore_all/NoOp

Help from Andre Hentz' via TensorFlow Lite <tflite@tensorflow.org>, I was told that frozen graph may have training ops.

Basically I would like to freeze the graph and convert it to quantized int8 tflite. If you need more details, could you please let me know?

[mnist.tar.gz](https://github.com/tensorflow/tensorflow/files/2142432/mnist.tar.gz)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by fzhu129** [comment]
not much related to environment but I can give my working environment:
1.  I don't write my custom code
2.  OS Platform and Distribution: Linux ubuntu 16.04
3.  git clone https://github.com/tensorflow/tensorflow.git to get latest tensorflow  (origin/master)
4.  bazel release 0.13.0
5. CUDA 9.1
6. GTX 1080 Ti, 11G
7  mnist from https://github.com/tensorflow/tensor2tensor

**Comment by fzhu129** [comment]
command to generate 
t2t-trainer   --generate_data   --data_dir=~/t2t_data   --output_dir=~/t2t_train/mnist   --problem=image_mnist   --model=shake_shake   --hparams_set=shake_shake_quick   --train_steps=1000   --eval_steps=100

/home/frankzhu/TestPrj/tensor2tensor/tensor2tensor/bin/t2t-trainer \
  --generate_data \
  --data_dir=~/t2t_data \
  --output_dir=/data/frankzhu/tmp/t2t_train/mnist \
  --problem=image_mnist \
  --model=shake_shake \
  --hparams_set=shake_shake_quick \
  --train_steps=1000 \
  --eval_steps=100 \
  --profile

**Comment by andrehentz** [comment]
Could you make your frozen graphdef available? From the message it looks like "save/restore_all/NoOp" is not an output of the model, but that might be misleading.

**Comment by fzhu129** [comment]
file size too big like  32225296 Jul  5 09:51 mnist.zip so not allow me to upload.

**Comment by fzhu129** [comment]
Hi Aselle:

If you still need that file, do you any place to upload large size file?
Or anything you want me to update from my side.

Thanks and best regards
Frank

**Comment by myth01** [comment]
@fzhu129 Please try "--output_array=save/restore_all/NoOp" instead of "--output_arrays=save/restore_all/NoOp"

**Comment by fzhu129** [comment]
@myth01 I tried this before and have tried today but not working.
bazel run --config=opt   //tensorflow/contrib/lite/toco:toco --   --input_file=/data/frankzhu/tmp/mnist.pb   --input_format=TENSORFLOW_GRAPHDEF   --output_format=TFLITE   --output_file=/data/frankzhu/tmp/mnist_q8.tflite   --inference_type=QUANTIZED_UINT8   --input_type=FLOAT   --input_arrays=input   --output_array=save/restore_all/NoOp --input_shapes=1,28,28,1 

The result is the same:
2018-07-17 11:54:12.054096: F tensorflow/contrib/lite/toco/tooling_util.cc:806] Check failed: model.HasArray(output_array) Output array not found: save/restore_all/NoOp

**Comment by myth01** [comment]
@fzhu129 In your freeze_graph command you use output_node_names=save/restore_shard_1. Did you try toco using the output node name instead of save/restore_all/NoOp

**Comment by fzhu129** [comment]
@myth01 have tried output_node_names=save/restore_shard_1 already and  it does not work.

**Comment by fzhu129** [comment]
Hi Suharsh:
Any update from your side?
Thanks 
Frank

**Comment by raghuraman-k** [comment]
Hi Frank,
 You cannot currently convert a training graph using TFLite. Note that TFLite only converts eval graphs via toco. For this purpose, you will need to create an eval mnist graph (without the optimizer)  as a pb and use it along with the checkpoint to call freeze_graph.

**Comment by fzhu129** [comment]
Hi Raghuraman:

for tensor2tensor, 

Do you have an example upon creating eval mnist graph (without the optimizer) as a pb with tensor2tensor?

So far I use this 
/home/frankzhu/TestPrj/tensor2tensor/tensor2tensor/bin/t2t-trainer 
--generate_data 
--data_dir=~/t2t_data 
--output_dir=/data/frankzhu/tmp/t2t_train/mnist 
--problem=image_mnist 
--model=shake_shake 
--hparams_set=shake_shake_quick 
--train_steps=1000 
--eval_steps=100 
--profile

how to evaluate it I have not found the command.

Thank you very much in advance
Frank

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
Hi I am not super familiar with tensor2tensor, but after a quick scan of their code, it seems they support saved model https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/bin/t2t_trainer.py#L171

This can then be converted with tflite_convert https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/python/tflite_convert.py

You should convert to FLOAT before trying quantized because currently you have to train models with Fakequant operations as described here: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/quantize which can be more complicated to start with.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tofulawrence** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
